### PR TITLE
Initial commit

### LINKED ISSUE #1: Implement an SQL dialect adapter for elastic-search

### PR DESCRIPTION
**Pull request by chiaradiamarcelo**
Closes #1

### CODE REVIEW

File: doc/changes/changelog.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Changes
+
+* [0.1.0](changes_0.1.0.md)
```
Comment (jakobbraun) [inline doc/changes/changelog.md:3]: file is missing

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.exasol.adapter.dialects.elasticsearch;
+
+import java.sql.Connection;
+
+import com.exasol.adapter.AdapterProperties;
+import com.exasol.adapter.dialects.*;
+import com.exasol.adapter.jdbc.*;
+
+public class ElasticSearchMetadataReader extends AbstractRemoteMetadataReader {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:9]: Add javadoc to all public and protected  classes, methds and interfaces. Only exception: visitors there you don't need javadoc for the visit methods

Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:9]: and for @override methods also not

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectFactory.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    public String getSqlDialectName() {
+        return ElasticSearchSqlDialect.NAME;
+    }
+
+    @Override
+    public String getSqlDialectVersion() {
+        return null;
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectFactory.java:22]: does not sound like a good idea

Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectFactory.java:22]: Check the other dialecs

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    public NullSorting getDefaultNullSorting() {
+        return NullSorting.NULLS_SORTED_HIGH;
+    }
+
+    @Override
+    protected RemoteMetadataReader createRemoteMetadataReader() {
+        try {
+            return new ElasticSearchMetadataReader(this.connectionFactory.getConnection(), this.properties);
+        } catch (final SQLException exception) {
+            throw new RemoteMetadataReaderException("Unable to create ElasticSearch remote metadata reader.",
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java:88]: Use error reporting + error codes

Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java:88]: This exception is not very meaningful. Is your meaningful message in the SQLException? In taht case it is ok. Otherwiese: Handle it earlier.

File: src/test/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+import com.exasol.adapter.AdapterProperties;
+import com.exasol.adapter.capabilities.Capabilities;
+import com.exasol.adapter.dialects.BaseQueryRewriter;
+import com.exasol.adapter.dialects.SqlDialect.NullSorting;
+import com.exasol.adapter.dialects.SqlDialect.StructureElementSupport;
+import com.exasol.adapter.jdbc.ConnectionFactory;
+
+@ExtendWith(MockitoExtension.class)
+public class ElasticSearchSqlDialectTest {
```
Comment (jakobbraun) [inline src/test/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectTest.java:31]: Don't forget to write integration tests later

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.elasticsearch.client</groupId>
+			<artifactId>elasticsearch-rest-high-level-client</artifactId>
+			<version>7.9.3</version>
+		</dependency>
+		<dependency>
+			<groupId>com.exasol</groupId>
+			<artifactId>exasol-testcontainers</artifactId>
+			<version>3.1.0</version>
```
Comment (jakobbraun) [inline pom.xml:98]: use 3.3.1 with Anasthasias fix
and then also testcontainer 1.15.0

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+		<dependency>
+			<groupId>org.testcontainers</groupId>
+			<artifactId>junit-jupiter</artifactId>
+			<version>${org.testcontainers.version}</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>com.exasol</groupId>
+			<artifactId>test-db-builder-java</artifactId>
+			<version>1.1.0</version>
```
Comment (jakobbraun) [inline pom.xml:110]: use 2.0.0

File: pom.xml [hunk 8]
```diff
@@ -0,0 +1,307 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (jakobbraun) [inline pom.xml:1]: Add content to your readme

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java (hunk 2 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:9]: and not for tests ;-)

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectTest.java (hunk 5 above)
Comment (chiaradiamarcelo) [reply src/test/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectTest.java:31]: yeah, this is gonna be in a following PR

File: pom.xml (hunk 8 above)
Comment (chiaradiamarcelo) [reply pom.xml:1]: this is going into another PR

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java (hunk 4 above)
Comment (chiaradiamarcelo) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java:88]: yes it is

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java [hunk 9]
```diff
@@ -6,6 +6,9 @@
 import com.exasol.adapter.dialects.*;
 import com.exasol.adapter.jdbc.*;
 
+/**
+ * Metadata reader implementation for reading metadata from an ElasticSearch remote database.
+ */
 public class ElasticSearchMetadataReader extends AbstractRemoteMetadataReader {
 
     public ElasticSearchMetadataReader(final Connection connection, final AdapterProperties properties) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:14]: add javadoc

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                .addPredicate(AND, OR, NOT, NOTEQUAL, LESS, LESSEQUAL, BETWEEN, IN_CONSTLIST, IS_NULL, IS_NOT_NULL,
+                        LIKE) //
+                .addLiteral(NULL, BOOL, DOUBLE, EXACTNUMERIC, STRING, INTERVAL) //
+                .addAggregateFunction(COUNT, COUNT_STAR, COUNT_DISTINCT, SUM, SUM_DISTINCT, MIN, MAX, AVG, AVG_DISTINCT,
+                        FIRST_VALUE, LAST_VALUE, STDDEV_POP, STDDEV_POP_DISTINCT, STDDEV_SAMP, STDDEV_SAMP_DISTINCT,
+                        VAR_POP, VAR_POP_DISTINCT, VAR_SAMP, VAR_SAMP_DISTINCT)//
+                .build();
+    }
+
+    public ElasticSearchSqlDialect(final ConnectionFactory connectionFactory, final AdapterProperties properties) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java:39]: add javadoc

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectFactory.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    public String getSqlDialectName() {
+        return ElasticSearchSqlDialect.NAME;
+    }
+
+    @Override
+    public String getSqlDialectVersion() {
+        return new VersionCollector("META-INF/maven/com.exasol/elasticsearch-virtual-schema/pom.properties")
+                .getVersionNumber();
+    }
+
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialectFactory.java:29]:
```suggestion
```

**Review by jakobbraun — CHANGES_REQUESTED** [review]
some small changs left

**Comment by jakobbraun** [comment]
Please fix the sonar findings

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import com.exasol.adapter.dialects.*;
+import com.exasol.adapter.jdbc.*;
+
+/**
+ * Metadata reader implementation for reading metadata from an ElasticSearch remote database.
+ */
+public class ElasticSearchMetadataReader extends AbstractRemoteMetadataReader {
+
+    /**
+     * Creates a new instance of {@link ElasticSearchMetadataReader}.
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:15]:
```suggestion
     * Create a new instance of {@link ElasticSearchMetadataReader}.
```

Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:15]: We use Create Build Get (without s)  to have the same style for all javadocs

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+                        LIKE) //
+                .addLiteral(NULL, BOOL, DOUBLE, EXACTNUMERIC, STRING, INTERVAL) //
+                .addAggregateFunction(COUNT, COUNT_STAR, COUNT_DISTINCT, SUM, SUM_DISTINCT, MIN, MAX, AVG, AVG_DISTINCT,
+                        FIRST_VALUE, LAST_VALUE, STDDEV_POP, STDDEV_POP_DISTINCT, STDDEV_SAMP, STDDEV_SAMP_DISTINCT,
+                        VAR_POP, VAR_POP_DISTINCT, VAR_SAMP, VAR_SAMP_DISTINCT)//
+                .build();
+    }
+
+    /**
+     * Creates a new instance of {@link ElasticSearchSqlDialect}.
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchSqlDialect.java:41]:
```suggestion
     * Createsa new instance of {@link ElasticSearchSqlDialect}.
```

**Review by jakobbraun — APPROVED** [review]
Approved with two suggestions

File: src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java (hunk 12 above)
Comment (chiaradiamarcelo) [reply src/main/java/com/exasol/adapter/dialects/elasticsearch/ElasticSearchMetadataReader.java:15]: The purpose of the javadoc in a method is to reply the question "What this method does?", which is always in present simple, as you can also see it as the standard way across the whole open jdk.

You could argue that one could write it in the infinitive form, as "To create" or "Creating", but just "Create" is not proper english.

**Pull request merged**